storage/watch.go

```diff
@@ -0,0 +1,333 @@
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	rev = s.KV.Put(key, value)
+	// TODO: avoid this range
+	kvs, _, err := s.KV.Range(key, nil, 0, rev)
+	if err != nil {
+		log.Panicf("unexpected range error (%v)", err)
+	}
+	s.notify(rev, storagepb.Event{
+		Type: storagepb.PUT,
+		Kv:   &kvs[0],
+	})
+	return rev
+}
+
+func (s *watchableStore) DeleteRange(key, end []byte) (n, rev int64) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// TODO: avoid this range
+	kvs, _, err := s.KV.Range(key, end, 0, 0)
+	if err != nil {
+		log.Panicf("unexpected range error (%v)", err)
+	}
+	n, rev = s.KV.DeleteRange(key, end)
+	for _, kv := range kvs {
+		s.notify(rev, storagepb.Event{
+			Type: storagepb.DELETE,
+			Kv: &storagepb.KeyValue{
+				Key: kv.Key,
+			},
+		})
+	}
+	return n, rev
+}
+
+func (s *watchableStore) TxnBegin() int64 {
+	s.mu.Lock()
+	s.tput = make(map[string]bool)
+	s.tdelete = make(map[string]bool)
+	return s.KV.TxnBegin()
+}
+
+func (s *watchableStore) TxnPut(txnID int64, key, value []byte) (rev int64, err error) {
+	rev, err = s.KV.TxnPut(txnID, key, value)
+	if err == nil {
+		s.tput[string(key)] = true
+		s.tdelete[string(key)] = false
+	}
+	return rev, err
+}
+
+func (s *watchableStore) TxnDeleteRange(txnID int64, key, end []byte) (n, rev int64, err error) {
+	kvs, _, err := s.KV.TxnRange(txnID, key, end, 0, 0)
+	if err != nil {
+		log.Panicf("unexpected range error (%v)", err)
+	}
+	n, rev, err = s.KV.TxnDeleteRange(txnID, key, end)
+	if err == nil {
+		for _, kv := range kvs {
+			s.tdelete[string(kv.Key)] = true
+			s.tput[string(kv.Key)] = false
+		}
+	}
+	return n, rev, err
+}
+
+func (s *watchableStore) TxnEnd(txnID int64) error {
+	err := s.KV.TxnEnd(txnID)
+	if err != nil {
+		return err
+	}
+
+	_, rev, _ := s.KV.Range(nil, nil, 0, 0)
+	for k := range s.tput {
+		kvs, _, err := s.KV.Range([]byte(k), nil, 0, 0)
+		if err != nil {
+			log.Panicf("unexpected range error (%v)", err)
+		}
+		s.notify(rev, storagepb.Event{
+			Type: storagepb.PUT,
+			Kv:   &kvs[0],
+		})
+	}
+	for k := range s.tdelete {
+		s.notify(rev, storagepb.Event{
+			Type: storagepb.DELETE,
+			Kv: &storagepb.KeyValue{
+				Key: []byte(k),
+			},
+		})
+	}
+	s.mu.Unlock()
+	return nil
+}
+
+func (s *watchableStore) Close() error {
+	close(s.stopc)
+	s.wg.Wait()
+	return s.KV.Close()
+}
+
+func (s *watchableStore) Watcher(key []byte, prefix bool, startRev, endRev int64) (Watcher, CancelFunc) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	wa := newWatcher(key, prefix, startRev, endRev)
+	k := string(key)
+	if startRev == 0 {
+		s.trackm[k] = append(s.trackm[k], wa)
+		if endRev != 0 {
+			s.endm[endRev] = append(s.endm[endRev], wa)
+		}
+	} else {
+		s.check = append(s.check, wa)
+	}
+
+	cancel := CancelFunc(func() {
+		s.mu.Lock()
+		s.mu.Unlock()
+		wa.stopWithError(ErrCanceled)
+
+		// remove global references of the watcher
+		for i, w := range s.check {
+			if w == wa {
+				s.check = append(s.check[:i], s.check[i+1:]...)
```

Comment:
yes. it may take time if there are lots of unsynced ones. It could be changed to a map in the future to avoid efforts.